File: python/bdi_client.py
from __future__ import annotations

import json
import os
import sys
import uuid
from pathlib import Path
from typing import Any

import requests

DEFAULT_HOST = "business-document-intelligence.p.rapidapi.com"
MAX_DOCUMENT_BYTES = 5 * 1024 * 1024
DEFAULT_TIMEOUT_SECONDS = 90


class BdiApiError(RuntimeError):
    """Raised when the API returns a non-success response."""


def _rapidapi_key() -> str:
    key = os.getenv("RAPIDAPI_KEY", "").strip()
    if not key:
        raise RuntimeError(
            "RAPIDAPI_KEY is not set. Set it in your environment before running this example."
        )
    return key


def _host() -> str:
    return os.getenv("RAPIDAPI_HOST", DEFAULT_HOST).strip() or DEFAULT_HOST


def _read_document(path_value: str) -> tuple[Path, bytes]:
    path = Path(path_value).expanduser().resolve()
    if not path.is_file():
        raise FileNotFoundError(f"Document not found: {path}")

    size = path.stat().st_size
    if size == 0:
        raise ValueError("Document is empty.")
    if size > MAX_DOCUMENT_BYTES:
        raise ValueError(
            f"Document is {size} bytes; the API maximum is {MAX_DOCUMENT_BYTES} bytes (5 MiB)."
        )

    return path, path.read_bytes()


def post_document(endpoint: str, document_path: str) -> dict[str, Any]:
    path, body = _read_document(document_path)
    host = _host()
    request_id = f"example.{uuid.uuid4().hex}"

    response = requests.post(
        f"https://{host}{endpoint}",
        headers={
            "Content-Type": "application/octet-stream",
            "X-RapidAPI-Host": host,
            "X-RapidAPI-Key": _rapidapi_key(),
            "X-Request-ID": request_id,
        },
        data=body,
        timeout=DEFAULT_TIMEOUT_SECONDS,
    )

    returned_request_id = response.headers.get("X-Request-ID")
    if returned_request_id:
        print(f"X-Request-ID: {returned_request_id}", file=sys.stderr)

    try:
        payload = response.json()
    except ValueError:
        payload = {"raw_response": response.text}

    if not response.ok:
        detail = json.dumps(payload, indent=2, sort_keys=True)
        raise BdiApiError(
            f"API request failed for {path.name} with HTTP {response.status_code}:\n{detail}"
        )

    if not isinstance(payload, dict):
        raise BdiApiError("API returned a successful response that was not a JSON object.")

    return payload


def print_json(payload: dict[str, Any]) -> None:
    print(json.dumps(payload, indent=2, sort_keys=True))

File: node/bdi-client.mjs
import { readFile, stat } from "node:fs/promises";
import { randomUUID } from "node:crypto";
import { resolve } from "node:path";

const DEFAULT_HOST = "business-document-intelligence.p.rapidapi.com";
const MAX_DOCUMENT_BYTES = 5 * 1024 * 1024;

function rapidApiKey() {
  const key = (process.env.RAPIDAPI_KEY ?? "").trim();
  if (!key) {
    throw new Error(
      "RAPIDAPI_KEY is not set. Set it in your environment before running this example."
    );
  }
  return key;
}

function rapidApiHost() {
  return (process.env.RAPIDAPI_HOST ?? DEFAULT_HOST).trim() || DEFAULT_HOST;
}

async function readDocument(pathValue) {
  const path = resolve(pathValue);
  const info = await stat(path);

  if (!info.isFile()) {
    throw new Error(`Document is not a file: ${path}`);
  }
  if (info.size === 0) {
    throw new Error("Document is empty.");
  }
  if (info.size > MAX_DOCUMENT_BYTES) {
    throw new Error(
      `Document is ${info.size} bytes; the API maximum is ${MAX_DOCUMENT_BYTES} bytes (5 MiB).`
    );
  }

  return { path, body: await readFile(path) };
}

export async function postDocument(endpoint, documentPath) {
  const { path, body } = await readDocument(documentPath);
  const host = rapidApiHost();
  const requestId = `example.${randomUUID()}`;

  const response = await fetch(`https://${host}${endpoint}`, {
    method: "POST",
    headers: {
      "content-type": "application/octet-stream",
      "x-rapidapi-host": host,
      "x-rapidapi-key": rapidApiKey(),
      "x-request-id": requestId,
    },
    body,
  });

  const returnedRequestId = response.headers.get("x-request-id");
  if (returnedRequestId) {
    console.error(`X-Request-ID: ${returnedRequestId}`);
  }

  const text = await response.text();
  let payload;
  try {
    payload = JSON.parse(text);
  } catch {
    payload = { raw_response: text };
  }

  if (!response.ok) {
    throw new Error(
      `API request failed for ${path} with HTTP ${response.status}:\n${JSON.stringify(payload, null, 2)}`
    );
  }

  if (payload === null || Array.isArray(payload) || typeof payload !== "object") {
    throw new Error("API returned a successful response that was not a JSON object.");
  }

  return payload;
}
